### ISSUE TITLE
depsolver concurrency limitations

### ISSUE THREAD

**Issue by matthiasr** [chef/chef-server#36]
(cross-filing as requested in opscode/chef-provisioning#112)

Using Open Source Chef Server 11.1.5 (but really ever since Chef 11) there seems to be a hard limit on the number of concurrent cookbook dependency resolutions.

We are running `chef-client` at regular intervals on all our hosts, but occasionally we want to kick them off immediately on one or more hosts, so we [pdsh](https://code.google.com/p/pdsh/) over them. This means that N (where N is the pdsh fanout) chef-client runs start at almost exactly the same time, and since they look up their cookbooks very early they nearly simultaneously hit the depsolvers.

Correlating with general chef-server load and the number of cookbooks / cookbook versions, it used to be that we could do this to no more than 5-10 nodes at the same time, in line with the default number of depsolvers (5). I bumped the `erchef['depsolver_worker_count']` to 20 now and can do this to ~30 nodes at once without hitting errors, but see errors at 40. Since they're probably not _perfectly_ in sync it looks to me like this is just hitting the new, higher limit again.

To sum up, it looks like each depsolver worker can only do one dependency resolution at a time, and it is not possible to have more than `depsolver_worker_count` simultaneous chef-client runs. Is that so, and is that by design or a known issue, or (quoting @jkeiser) "bad mojo"?

**Comment by sdelano** [comment]
Coming back to this issue again, the conclusion still stands that we should add queuing to the depsolver workers. I'll add the `accepted-minor` tag to this card, indicating that we should do this and that it's a minor issue.

We should also include upgrading to the newest `dep_selector` gem if we haven't done so already. @danielsdeleo made some great improvements there to optimize the Ruby portion of the code so that it doesn't have to do as much work building up redundant objects.

**Comment by heaven** [comment]
Hi, why is it a minor issue? We have 25 nodes and they all constantly fail to run chef-client.

**Comment by matthiasr** [comment]
You can work around it by raising the number of dependency solvers. The
default is very conservative.

**Comment by heaven** [comment]
Yeah, thank you, looking into that now.
